### ISSUE TITLE
For of + switch behaves incorrectly

### ISSUE THREAD

**Issue by Y0ba** [babel/babel#963]
Code:
```js
function* gen() {
  yield 1;
  yield 2;
  yield 3;
}

function forOfSwitch() {
  let rv = 0;
  for(let a of gen()) {
    switch(a) {
      case 1: rv++; break;
      case 2: rv++; break;
      case 3: return rv;
    }
  }
}

function forOfIf() {
  let rv = 0;
  for(let a of gen()) {
    if(a === 1) {
      rv++;
    } else if(a === 2) {
      rv++;
    } else if(a === 3) {
      return rv;
    }
  }
}

alert(forOfSwitch() === forOfIf());
```

**Expected result**: it should alert `true`.
**Actual result**: alerts `false`.

**Comment by Y0ba** [comment]
Also, related problem. Code:
```js
function Iterable() {
  this.data = [1,2,3];
}
Iterable.prototype = {
  *[Symbol.iterator]() {
    yield* this.data;
  }
}

function forOfSwitch() {
  let rv = 0;
  for(let a of new Iterable()) {
    switch(a) {
      case 1: rv++; break;
      case 2: rv++; break;
      case 3: return rv;
    }
  }
}

alert(forOfSwitch());
```
**Expected result**: it should alert `2`.
**Actual result**: `TypeError: fn is undefined`.

**Comment by sebmck** [comment]
Probably an issue with transforming the `break`s inside the case statements.

**Comment by Y0ba** [comment]
Second testcase throws an error even without `break`s:
```js
function Iterable() {
  this.data = [1,2,3];
}
Iterable.prototype = {
  *[Symbol.iterator]() {
    yield* this.data;
  }
}

function foo() {
  for(let a of new Iterable()) {
    return a;
  }
}

alert(foo());
```

**Comment by sebmck** [comment]
@Y0ba I need more information, can you provide the stacktrace? Not much I can debug without it :stuck_out_tongue_winking_eye:

**Comment by sebmck** [comment]
Fixed as of 4.7.2, thanks! Both of your examples work fine now.
